File: backend/app/models/book.py
from datetime import datetime, timezone

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.tag import book_tags


class Book(Base):
    __tablename__ = "books"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    author: Mapped[str] = mapped_column(String(150), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    category_id: Mapped[int | None] = mapped_column(ForeignKey("categories.id", ondelete="SET NULL"), nullable=True)
    cloudinary_url: Mapped[str] = mapped_column(String(500), nullable=False)
    cloudinary_public_id: Mapped[str] = mapped_column(String(255), nullable=False)
    cover_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    total_pages: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    category = relationship("Category", back_populates="books")
    tags = relationship("Tag", secondary=book_tags, back_populates="books")
    bookmarks = relationship("Bookmark", back_populates="book", cascade="all, delete-orphan")
    notes = relationship("Note", back_populates="book", cascade="all, delete-orphan")

File: frontend/src/types/book.ts
export interface Category {
  id: number;
  name: string;
  slug: string;
}

export interface Tag {
  id: number;
  name: string;
}

export interface Book {
  id: number;
  title: string;
  author: string;
  description: string | null;
  cover_url: string | null;
  total_pages: number | null;
  created_at: string;
  category: Category | null;
  tags: Tag[];
}

export interface BookDetail extends Book {
  cloudinary_url: string;
}

export interface BooksResponse {
  items: Book[];
  total: number;
  page: number;
  page_size: number;
}

export interface Bookmark {
  id: number;
  book_id: number;
  page_number: number;
  label: string | null;
  created_at: string;
}

export interface Note {
  id: number;
  book_id: number;
  page_number: number;
  text: string;
  created_at: string;
}
